File: backend/apps/quotations/schemas.py
from datetime import date, datetime
from decimal import Decimal
from typing import Any
from uuid import UUID

from ninja import Schema
from pydantic import EmailStr, Field


class QuotationItemCreateSchema(Schema):
    title: str = Field(min_length=1, max_length=250)
    description: str = ""
    quantity: Decimal = Field(default=Decimal("1.00"), gt=0)
    unit_price: Decimal = Field(default=Decimal("0.00"), ge=0)
    discount_amount: Decimal = Field(default=Decimal("0.00"), ge=0)
    tax_rate: Decimal = Field(default=Decimal("0.00"), ge=0)
    sort_order: int = Field(default=0, ge=0)


class QuotationItemSchema(Schema):
    id: UUID
    title: str
    description: str
    quantity: Decimal
    unit_price: Decimal
    discount_amount: Decimal
    tax_rate: Decimal
    subtotal: Decimal
    tax_amount: Decimal
    total_amount: Decimal
    sort_order: int


class QuotationRecipientCreateSchema(Schema):
    name: str = Field(default="", max_length=200)
    email: EmailStr
    is_primary: bool = False


class QuotationRecipientSchema(Schema):
    id: UUID
    name: str
    email: str
    is_primary: bool
    received_at: datetime | None
    viewed_at: datetime | None


class QuotationEventSchema(Schema):
    id: UUID
    event_type: str
    description: str
    metadata: dict[str, Any]
    created_at: datetime


class QuotationSchema(Schema):
    id: UUID
    quotation_number: str
    client_id: UUID
    client_name: str
    lead_id: UUID | None
    title: str
    subject: str
    description: str
    status: str
    issue_date: date
    expiry_date: date | None
    currency: str
    subtotal: Decimal
    discount_amount: Decimal
    tax_amount: Decimal
    total_amount: Decimal
    terms: str
    notes: str
    accepted_at: datetime | None
    accepted_by_name: str
    accepted_by_email: str
    sent_at: datetime | None
    duplicated_from_id: UUID | None
    is_expired: bool
    items: list[QuotationItemSchema]
    recipients: list[QuotationRecipientSchema]
    events: list[QuotationEventSchema]
    created_at: datetime
    updated_at: datetime


class QuotationCreateSchema(Schema):
    client_id: UUID
    lead_id: UUID | None = None
    title: str = Field(min_length=1, max_length=250)
    subject: str = Field(default="", max_length=250)
    description: str = ""
    issue_date: date | None = None
    expiry_date: date | None = None
    currency: str = Field(default="LKR", min_length=3, max_length=3)
    discount_amount: Decimal = Field(default=Decimal("0.00"), ge=0)
    tax_amount: Decimal = Field(default=Decimal("0.00"), ge=0)
    terms: str = ""
    notes: str = ""
    items: list[QuotationItemCreateSchema] = Field(default_factory=list)
    recipients: list[QuotationRecipientCreateSchema] = Field(
        default_factory=list
    )


class QuotationUpdateSchema(Schema):
    title: str = Field(min_length=1, max_length=250)
    subject: str = Field(default="", max_length=250)
    description: str = ""
    issue_date: date
    expiry_date: date | None = None
    currency: str = Field(min_length=3, max_length=3)
    discount_amount: Decimal = Field(default=Decimal("0.00"), ge=0)
    tax_amount: Decimal = Field(default=Decimal("0.00"), ge=0)
    terms: str = ""
    notes: str = ""
    items: list[QuotationItemCreateSchema] = Field(default_factory=list)
    recipients: list[QuotationRecipientCreateSchema] = Field(
        default_factory=list
    )


class QuotationAcceptSchema(Schema):
    accepted_by_name: str = Field(min_length=1, max_length=200)
    accepted_by_email: EmailStr


class QuotationRejectSchema(Schema):
    reason: str = ""


class QuotationPdfPayloadSchema(Schema):
    company: dict[str, Any]
    client: dict[str, Any]
    quotation: dict[str, Any]
    items: list[dict[str, Any]]
    totals: dict[str, Any]

File: apps/dashboard/src/features/quotations/schemas.ts
import {
  z,
} from "zod";

import {
  quotationStatuses,
} from "./types";

const quotationItemSchema = z.object({
  id: z.string().uuid(),
  title: z.string(),
  description: z.string(),
  quantity: z.string(),
  unit_price: z.string(),
  discount_amount: z.string(),
  tax_rate: z.string(),
  subtotal: z.string(),
  tax_amount: z.string(),
  total_amount: z.string(),
  sort_order: z.number().int().nonnegative(),
});

const quotationRecipientSchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  email: z.string(),
  is_primary: z.boolean(),
  received_at: z.string().nullable(),
  viewed_at: z.string().nullable(),
});

const quotationEventSchema = z.object({
  id: z.string().uuid(),
  event_type: z.string(),
  description: z.string(),
  metadata: z.record(
    z.string(),
    z.unknown(),
  ),
  created_at: z.string(),
});

export const quotationSchema = z.object({
  id: z.string().uuid(),
  quotation_number: z.string(),
  client_id: z.string().uuid(),
  client_name: z.string(),
  lead_id: z.string().uuid().nullable(),
  title: z.string(),
  subject: z.string(),
  description: z.string(),
  status: z.enum(quotationStatuses),
  issue_date: z.string(),
  expiry_date: z.string().nullable(),
  currency: z.string(),
  subtotal: z.string(),
  discount_amount: z.string(),
  tax_amount: z.string(),
  total_amount: z.string(),
  terms: z.string(),
  notes: z.string(),
  accepted_at: z.string().nullable(),
  accepted_by_name: z.string(),
  accepted_by_email: z.string(),
  sent_at: z.string().nullable(),
  duplicated_from_id: z.string().uuid().nullable(),
  is_expired: z.boolean(),
  items: z.array(quotationItemSchema),
  recipients: z.array(
    quotationRecipientSchema,
  ),
  events: z.array(quotationEventSchema),
  created_at: z.string(),
  updated_at: z.string(),
});

export const paginatedQuotationsSchema =
  z.object({
    items: z.array(quotationSchema),
    pagination: z.object({
      page: z.number().int().positive(),
      page_size: z.number().int().positive(),
      total_items: z.number().int().nonnegative(),
      total_pages: z.number().int().nonnegative(),
    }),
  });
